Hantec API Ecommerce

JAntonioSalas opened the pull request:
Revisión de código

JAntonioSalas committed:
[IMP] various: add and enhance several endpoints

Added docstring, clarified the purpose and improved of multiple endpoints to enhance the user experience and maintain consistency in the application.

1. `/create_schedule_activity_invoice`: Improved error handling for creating scheduled activities related to invoices.
2. `/get_states_mexico`: Enhanced the endpoint to retrieve the list of states in Mexico, including better error handling and more detailed responses.
3. `/get_inventory_by_sku`: Updated the endpoint to retrieve inventory details based on SKU, ensuring accurate and detailed inventory information is returned.
4. `/get_inventory`: Improved the retrieval of inventory details for all products with a SKU, adding more detailed responses.
5. `/send_message_sale_order`: Added functionality to post messages to sale orders, with better logging and error handling.
6. `/create_schedule_activity`: Enhanced the creation of scheduled activities related to sale orders, including additional validation and detailed responses.
7. `/confirm_sale_order`: Simplified the process of confirming sale orders and added detailed logging and responses.
8. `/send_invoice_by_email/<int:invoice_id>`: Improved the functionality for sending invoices by email, ensuring better user feedback and error handling.
9. `/stamp_invoice/<int:invoice_id>`: Enhanced the invoice stamping process, including detailed error messages and success responses.
10. `/download_invoice/<int:invoice_id>`: Streamlined the invoice download process to generate and deliver PDFs more efficiently.
11. `/get_shipping_info/<model("sale.order"):order>`: Added functionality to retrieve detailed shipping information for sale orders, including delivery address details and shipping records.
12. `/register_payment_invoice/<model("account.move"):invoice>`: Improved the payment registration process for invoices, ensuring accurate logging and error handling.
13. `/invoice_sale_order`: Enhanced the process of invoicing sale orders, including better handling of invoice creation and posting.
14. `/update_sale_order`: Improved the update process for sale orders, including validation and detailed logging.
15. `/create_sale_order`: Enhanced the creation process for sale orders with better validation and handling of optional fields.
16. `/delivery_address`: Improved the process for creating or updating delivery addresses, including better handling of address formatting and validation.
17. `/address_invoice`: Enhanced the process for creating or updating billing addresses, ensuring better validation and error handling.
18. `/update_contact`: Improved the contact update process, including better validation and detailed logging.
19. `/create_contact`: Enhanced the contact creation process with better search logic and validation based on email, phone, or store name (contact name).

These changes improve the overall robustness and maintainability of the system by ensuring that each endpoint handles data consistently and provides meaningful feedback to users.

Closes #24  (close related PR on GitHub)

hugho-ad commented on hantec_api_ecommerce/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import controllers
```
EOF

leave and empty line at the end of file

apply it to all files please

hugho-ad commented on hantec_api_ecommerce/__manifest__.py:
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "Hantec API Ecommerce",
+    "summary": "Hantece API Ecommerce",
+    "version": "1.0.0",
```
```suggestion
    "version": "15.0.0.0",
```

hugho-ad commented on hantec_api_ecommerce/__manifest__.py:
```diff
@@ -0,0 +1,12 @@
+
```
```suggestion
```

hugho-ad commented on hantec_api_ecommerce/controllers/__init__.py:
```diff
@@ -0,0 +1,3 @@
+
```
```suggestion
```

hugho-ad commented on hantec_api_ecommerce/controllers/__init__.py:
```diff
@@ -0,0 +1,3 @@
+
+
```
```suggestion
```

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+
+from odoo.http import request, Response, Controller, route
```
```suggestion
```
move it after import logging

hugho-ad replied:
Response us not used

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+
+from odoo.http import request, Response, Controller, route
+import json
+import logging
```
```suggestion
import logging
from odoo.http import Controller, request,  route
```

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+
+
+class MainController(Controller):
+    @route('/create_contact', methods=['POST'], type='json', auth='user' )
```
```suggestion
    @route("/create_contact", methods=["POST"], type="json", auth="user")
```
Use double quoting please, check all coincidences

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        # Lista de todos los campos posibles
+
+        all_fields = [
+            'company_type', 'name', 'phone', 'street_number', 'street_number2',
```
keep as a list 
```suggestion
            'company_type', 
            'name', 
            'phone', 
            'street_number', 
            'street_number2',
```

an so on, replace single quote by double quoting

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+            'vat', 'l10n_mx_edi_curp', 'website', 'lang', 'category_id', 'l10n_mx_edi_fiscal_regime', 'parent_id'
+        ]
+
+        # Construir el diccionario contact_data con solo los campos enviados en el JSON
```
We prefer English 
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#idioms

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        ]
+
+        # Construir el diccionario contact_data con solo los campos enviados en el JSON
+        contact_data = {field: request.jsonrequest.get(field) for field in all_fields if request.jsonrequest.get(field) is not None}
```
```suggestion
        contact_data = {
            field: request.jsonrequest.get(field) for field in all_fields if request.jsonrequest.get(field) is not None
        }
```

take care of indentation please

you can use black 
https://pypi.org/project/black/

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+            existing_contact.write(contact_data)
+            logger.info("Contacto actualizado con ID %s", existing_contact.id)
+            return {"message": f"Contacto actualizado con ID: {existing_contact.id}.", "contact_id": existing_contact.id}
+        else:
```
```suggestion
```
 Unnecessary "else" after "return", remove the "else" and de-indent the code inside it

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+
+            logger.info("Contacto encontrado con ID %s", existing_contact.id)
+            return {"message": f"Contacto encontrado con ID: {existing_contact.id}.", "contact_id": existing_contact.id}
+        else:
```
Unnecessary "else" after "return", remove the "else" and de-indent the code inside it

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        # Verificar nuevamente si la factura fue timbrada exitosamente
+        if invoice.l10n_mx_edi_cfdi_uuid:
+            return {"success": "La factura ha sido timbrada con éxito.", "UUID": invoice.l10n_mx_edi_cfdi_uuid}
+        else:
```
Unnecessary "else" after "return", remove the "else" and de-indent the code inside it

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+            "sales_team": order.team_id.id,
+            "market_place_reference": order.channel_order_reference,
+            "sale_order_name": order.name,
+            "picking_id": picking.id,
```
Using possibly undefined loop variable 'picking'

picking is defined as a for loop var only

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+            "market_place_reference": order.channel_order_reference,
+            "sale_order_name": order.name,
+            "picking_id": picking.id,
+            "picking_name": picking.name
```
same

hugho-ad commented on hantec_api_ecommerce/__manifest__.py:
```diff
@@ -0,0 +1,12 @@
+    "summary": "Hantece API Ecommerce",
+    "version": "1.0.0",
+    "author": "C&O PROJECTS AND SOLUTIONS",
+    "category": "Sales Managment",
```
Missing required key "license" in manifest file

hugho-ad commented:
Missing ./README.rst file. Template here: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+            'active_id': order.id,
+        }
+        
+        invoice_wizard = request.env['sale.advance.payment.inv'].with_context(context).create({
```
```suggestion
        invoice_wizard = request.env['sale.advance.payment.inv'].with_context(**context).create({
```

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        contact_data = {field: request.jsonrequest.get(field) for field in all_fields if request.jsonrequest.get(field) is not None}
+
+        # Si category_id está presente, convertirlo en una lista
+        if 'category_id' in contact_data and not isinstance(contact_data['category_id'], list):
```
because category_id is a m2m field?

which error do you get if you pass the id right as it is?

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+
+        for key, value in contact_data.items():
+            if isinstance(value, str) and key not in list_not_title:
+                    contact_data[key] = value.title()
```
why did you change the values?

you should register as it comes, not data transforming

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+
+        list_not_title = ['lang', 'company_type', 'email', 'website', 'vat']
+
+        for key, value in update_vals.items():
```
repeted code, should be this a function

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        contact_data = data.get('contact_data', {})  # Datos del contacto a actualizar/crear
+
+        # Verificar si ya existe un contacto a partir de los primeros digitos del correo y ultimos digitos del telefono
+        email_prefix = email[0:5]
```
better apply a regex to get the email before the domain and the @

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        phone_suffix = phone[len(phone)-5:len(phone)]
+
+        # Crear dominios de búsqueda usando patrones con comodines
+        domain = [('email', 'ilike', f'{email_prefix}%'), ('phone', 'ilike', f'%{phone_suffix}')]
```
why no you use the ids?

I mean, you site are creating contacts and returning it's ids to allow post identification

JAntonioSalas replied:
Because there are already contacts in odoo that we want to know if they exist before creating them

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        return {"message": f"Contacto con ID: {partner_id} actualizado exitosamente."}
+    
+    @route('/manage_contact', methods=['POST'], type='json', auth='user')
+    def manage_contact(self):
```
each endpoint has to have a method docstring please

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        domain = [('name', 'ilike', f'%{store_name_suffix}')]
+
+        # Realizar la búsqueda con el dominio definido
+        existing_contact = request.env['res.partner'].search(domain, limit=1)
```
seams duplicated endpoint, just by the store_name

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,570 @@
+        invoice_data = data.get('invoice_data')
+
+        partner = env['res.partner'].browse(partner_id)
+
```
avoid to many blank lines please,

JAntonioSalas committed:
[FIX] base: correct response type for get_states_mexico endpoint

Changed the response type from 'http' to 'json' for the /get_states_mexico endpoint
to ensure the correct content type is returned and improve API consistency.

The previous configuration had the type set to 'http', which caused issues with
clients expecting a JSON response. This fix ensures that the endpoint now returns
a JSON response, aligning with the expected behavior and improving the overall
reliability of the API.

JAntonioSalas committed:
[FIX] base: correct phone suffix length in create_contact endpoint

Changed the phone suffix length from 4 to 5 in the create_contact endpoint to ensure
the correct portion of the phone number is used for matching existing contacts.

The previous implementation used the last 4 digits of the phone number, which was insufficient
for accurate matching. Updating this to use the last 5 digits improves the reliability
of contact searches and prevents mismatches.

JAntonioSalas committed:
[IMP] various: clean up whitespace and update manifest

Cleaned up whitespace in initialization files and updated the module manifest.

- Removed unnecessary whitespace from `__init__.py` files to improve readability and maintain coding standards.
- Updated the `__manifest__.py` to include the new version and added the license field for compliance and clarity.

These changes enhance the code quality and ensure the module metadata is up-to-date and correctly specified.

Closes #24  (close related PR on GitHub)

JAntonioSalas committed:
[ADD] doc: add README.rst for module documentation

Added a README.rst file to provide comprehensive documentation for the module.
This README includes:

- A summary of the module's functionality.
- Information about authors, maintainers, development status, and license.
- A changelog section detailing recent updates and features.

This documentation aims to improve the understanding and usability of the module for all users and contributors.

Closes #24  (close related PR on GitHub)
